Spinweight

moxcodes opened the pull request:
## Proposed changes

Add spin-weight wrapper and convenience functions for random values used in the spin-weighted tests. The spin-weight wrapper ensures that math operations are permitted only for valid combinations of spin-weight. Unwrapped objects are treated as spin-weight 0 (this might be a concerning choice; it is definitely looser than insisting that spin-weights are always consistent).

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Add convenience macro for seeding a generator : MAKE_GENERATOR(NAME)

moxcodes committed:
Add SpinWeighted wrapper

- SpinWeighted wraps any type and constrains operations to obey spin-weight
  rules.

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,80 @@
+/// spins are the same.
+template <typename T1, typename T2, int Spin,
+          typename Ret = decltype(std::declval<T1>() + std::declval<T2>())>
+SPECTRE_ALWAYS_INLINE SpinWeighted<Ret, Spin> operator+(
```
What you can do to set the return type without a template parameter is:
```
SPECTRE_ALWAYS_INLINE auto operator+(const SpinWeighted<T1, Spin>& lhs,
                        const SpinWeighted<T2, Spin>& rhs) noexcept
                    -> SpinWeighted<decltype(std::declval<T1>() + std::declval<T2>()), Spin> {
```
You can do the same to all the ones below :)

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,80 @@
+/// \brief Test equivalence of spin-weighted quantities if the types are
+/// compatible and spins are the same.
+template <typename T1, typename T2, int Spin>
+SPECTRE_ALWAYS_INLINE bool operator==(
```
Could you add `operator!=` too? It can just do `return not(lhs == rhs);` (no data, so this way if `operator==` were to be updated, `!=` gets it for free)

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,80 @@
+  constexpr static int spin = Spin;
+};
+
+// @{
```
You don't need the `// @{` and `// @}` since you only have once function the Doxygen comment is being applied to. The `// @{` means "Apply the first Doxygen comment to all functions inside the `// @{` and `// @}`."

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,80 @@
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief A wrapper for a spin-weighting wrapper which limits math operations to
```
I find this sentence difficult to parse. Maybe something like:

  Make a spin-weighted type `T` with spin-weight `Spin`. Mathematical operators are restricted to addition, subtraction, multiplication and division.

would be clearer?

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,46 @@
+  auto gen = make_generator();
+  UniformCustomDistribution<double> dist{-100.0, 100.0};
+
+  auto v1 = make_with_random_values<SpinWeighted<double, 0>>(
```
I think all these can be `const`.

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,46 @@
+#include "tests/Utilities/MakeWithRandomValues.hpp"
+
+SPECTRE_TEST_CASE("Unit.DataStructures.SpinWeight", "[DataStructures][Unit]") {
+  auto gen = make_generator();
```
Could you factor out the operations an just `double` and template on the type, then do double, int and long?

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,46 @@
+      make_not_null(&gen), make_not_null(&dist));
+  auto v2 = make_with_random_values<SpinWeighted<double, 1>>(
+      make_not_null(&gen), make_not_null(&dist));
+  auto v3 = make_with_random_values<SpinWeighted<double, 2>>(
```
Could you add one with a negative spin-weight?

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,46 @@
+  CHECK(v5 / v3 == SpinWeighted<std::complex<double>, -1>{v5.data / v3.data});
+
+  int spin = decltype(v1)::spin;
+  CHECK(spin == 0);
```
The spin checks should be `static_assert` to verify that it could be evaluated at compile time.

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -27,54 +27,103 @@ struct SpinWeighted {
 
```
The summary line in the commit message is a bit long

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -27,54 +27,103 @@ struct SpinWeighted {
   return {lhs.data + rhs.data};
 }
+template <typename T>
+SPECTRE_ALWAYS_INLINE SpinWeighted<T, 0> operator+
```
The formatting looks off here

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -27,54 +27,103 @@ struct SpinWeighted {
   return lhs.data == rhs.data;
 }
+template <typename T>
+SPECTRE_ALWAYS_INLINE bool operator==(SpinWeighted<T, 0> lhs, T rhs) noexcept {
```
Can you add `operator!=` for the two new ones?

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -44,3 +54,29 @@ SPECTRE_TEST_CASE("Unit.DataStructures.SpinWeight", "[DataStructures][Unit]") {
   CHECK(spin == 2);
 }
+
+/// \cond HIDDEN_SYMBOLS
```
Can you put the classes in an anonymous namespace?

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -27,54 +27,103 @@ struct SpinWeighted {
+}
+template <typename T>
+SPECTRE_ALWAYS_INLINE SpinWeighted<T, 0> operator+
+(T lhs, SpinWeighted<T,0> rhs) noexcept {
```
It's standard to use `const T& lhs, const SpinWeighted<T, 0>& rhs` (const references).

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,80 @@
+template <typename T1, typename T2, int Spin,
+          typename Ret = decltype(std::declval<T1>() - std::declval<T2>())>
+SPECTRE_ALWAYS_INLINE SpinWeighted<Ret, Spin> operator-(
+    SpinWeighted<T1, Spin> lhs, SpinWeighted<T2, Spin> rhs) noexcept {
```
It's standard to use `const SpinWeighted<T1, Spin>& lhs, const SpinWeighted<T2, Spin>& rhs` (const &) for these

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -44,3 +54,29 @@ SPECTRE_TEST_CASE("Unit.DataStructures.SpinWeight", "[DataStructures][Unit]") {
+      cpp17::void_t<decltype(std::declval<T1>() OP std::declval<T2>())>> \
+      : std::false_type {};                                              \
+  static_assert(TAG<LHSTYPE, RHSTYPE>::value, /*NOLINT*/                \
+                "static_assert failed, " #LHSTYPE #OP #RHSTYPE " had a type");
```
remove `;`

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -44,3 +54,29 @@ SPECTRE_TEST_CASE("Unit.DataStructures.SpinWeight", "[DataStructures][Unit]") {
+CHECK_TYPE_OPERATION_FAIL(spin_check_2, +, SpinOne, SpinTwo);
+CHECK_TYPE_OPERATION_FAIL(spin_check_3, +,
+                          decltype(std::declval<SpinZero>() *
+                                   std::declval<SpinTwo>()), SpinOne)
```
add `;`

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the changes

moxcodes committed:
Treat unweighted values as spin 0; more tests

- Add treatment of wrapped types as implicitly of Spin-0.
- Add test that invalid spin-weight operations fails to compile (via SFINAE)

moxcodes commented:
Fixups squashed in.
Cheers!

nilsdeppe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,96 @@
+    test_spinweights<typename decltype(x)::type>();
+  });
+}
+
```
only one blank line

nilsdeppe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,159 @@
+
+/*!
+ * \ingroup DataStructuresGroup
+ * Make a spin-weighted type `T` with spin-weight `Spin`. Mathematical operators
```
You'll need a `\brief` at the beginning of this line.

nilsdeppe reviewed (commented):
Couple minor things I'm again happy for you to squash :)

moxcodes commented:
Minor fixups squashed in.
Cheers!

moxcodes commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,159 @@
+ * pair of spin-weighted quantities - i.e. addition only makes sense when the
+ * two summands possess the same spin weight, and multiplication (or division)
+ * result in a summed (or subtracted) spin weight.
+ *
```
remove

moxcodes commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,159 @@
+/// add the spins. Un-weighted quantities are assumed to be spin 0.
+template <typename T1, typename T2, int Spin1, int Spin2>
+SPECTRE_ALWAYS_INLINE SpinWeighted<
+    decltype(std::declval<T1>() + std::declval<T2>()), Spin1 + Spin2>
```
+->* (oops)

moxcodes commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,159 @@
+/// spins are the same. Un-weighted quantities are assumed to be spin 0.
+template <typename T1, typename T2, int Spin>
+SPECTRE_ALWAYS_INLINE
+    SpinWeighted<decltype(std::declval<T1>() + std::declval<T2>()), Spin>
```
+->- (oops)

moxcodes commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,159 @@
+/// subtract the spins. Un-weighted quantities are assumed to be spin 0.
+template <typename T1, typename T2, int Spin1, int Spin2>
+SPECTRE_ALWAYS_INLINE SpinWeighted<
+    decltype(std::declval<T1>() + std::declval<T2>()), Spin1 - Spin2>
```
again ops

moxcodes reviewed (commented):
group code review with @markscheel @fmahebert @mgiesler

fmahebert commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,95 @@
+  CHECK(v2 * v3 == SpinWeighted<T, -1>{v2.data * v3.data});
+  CHECK(v5 / v3 == SpinWeighted<std::complex<T>, 3>{v5.data / v3.data});
+
+  // check the plain data types act as spin 0
```
the -> that

fmahebert reviewed (commented):
A small word fix, if you haven't already done the other changes

moxcodes commented:
Fixups squashed.
Cheers!

nilsdeppe reviewed (dismissed):
Before you try to write a special vector for this we should talk because I don't think that'll be necessary. The way you've coded things here I think just beefing up the SpinWeighted class a bit in a separate PR will allow you to do `SpinWeighted<DataVector, SPIN>` and have everything work the way you want.

kidder commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -26,60 +25,134 @@ struct SpinWeighted {
 template <typename T1, typename T2, int Spin1, int Spin2>
 SPECTRE_ALWAYS_INLINE SpinWeighted<
-    decltype(std::declval<T1>() + std::declval<T2>()), Spin1 - Spin2>
+    decltype(std::declval<T1>() / std::declval<T2>()), Spin1 - Spin2>
```
this (and the similar fixes) should be in previous commit, or the last two commits should be squashed into one

moxcodes commented:
Changes re-allocated to appropriate commits.
Cheers!

nilsdeppe reviewed (dismissed):
@kidder @wthrowe could you please look at this when you get a chance? Thanks! :)

wthrowe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -336,3 +337,16 @@ void test_throw_exception(const ThrowingFunctor& func,
+GenType make_generator() noexcept {
+  std::random_device r;
+  const auto seed = r();
+  INFO("Seed is: " << seed);
```
This won't work.  INFO creates an RAII-style object, so it needs to be in a scope enclosing the tests.

wthrowe commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -0,0 +1,95 @@
+                                            make_not_null(&dist));
+  const auto v8 = make_with_random_values<std::complex<T>>(make_not_null(&gen),
+                                                          make_not_null(&dist));
+  // check compile-timee spin values
```
`time`

wthrowe commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -0,0 +1,158 @@
+ * pair of spin-weighted quantities - i.e. addition only makes sense when the
+ * two summands possess the same spin weight, and multiplication (or division)
+ * result in a summed (or subtracted) spin weight.
+ */
```
There is a change here that is in the wrong commit.

nilsdeppe reviewed (dismissed):
@kidder @wthrowe could you please look at this when you get a chance? Thanks!

wthrowe commented:
Update your commit message.

moxcodes commented:
commit message updated for the `MAKE_GENERATOR` macro.
Thanks!

nilsdeppe reviewed (approved):
@kidder
